lib/jsdom/browser/htmltodom.js

```diff
@@ -67,90 +67,89 @@ module.exports = class HTMLToDOM {
     return contextNode;
   }
 
-  _parseWithSax(html, isFragment, contextNode) {
-    const SaxParser = this.parser.parser;
-    const parser = new SaxParser(/* strict = */true, { xmlns: true, strictEntities: true });
-    parser.noscript = false;
-    parser.looseCase = "toString";
+  _parseWithSaxes(html, isFragment, contextNode) {
+    const parserOptions = { xmlns: true };
+    if (isFragment) {
+      parserOptions.fragment = true;
+
+      parserOptions.resolvePrefix = prefix => {
+        // saxes wants undefined as the return value if the prefix is not
+        // defined, not null.
+        return contextNode.lookupNamespaceURI(prefix) || undefined;
+      };
+    }
+    const parser = new this.parser.SaxesParser(parserOptions);
     const openStack = [contextNode];
+    const currentDocument = contextNode._ownerDocument || contextNode;
+
     parser.ontext = text => {
-      setChildForSax(openStack[openStack.length - 1], {
-        type: "text",
-        data: text
-      });
+      appendChild(
+        openStack[openStack.length - 1],
+        currentDocument.createTextNode(text)
+      );
     };
     parser.oncdata = cdata => {
-      setChildForSax(openStack[openStack.length - 1], {
-        type: "cdata",
-        data: cdata
-      });
+      appendChild(
+        openStack[openStack.length - 1],
+        currentDocument.createCDATASection(cdata)
+      );
     };
-    parser.onopentag = arg => {
-      const attrs = Object.keys(arg.attributes).map(key => {
-        const rawAttribute = arg.attributes[key];
+    parser.onopentag = tag => {
+      const { local: tagLocal, uri: tagURI, prefix: tagPrefix } = tag;
+      const elem = currentDocument._createElementWithCorrectElementInterface(tagLocal, tagURI);
+      elem._prefix = tagPrefix || null;
+      elem._namespaceURI = tagURI || null;
+      // We mark a script element as "parser-inserted", which prevents it from
+      // being immediately executed.
+      if (tagLocal === "script" && tagURI === HTML_NS) {
+        elem._parserInserted = true;
+      }
 
-        let { prefix } = rawAttribute;
-        let localName = rawAttribute.local;
-        if (prefix === "xmlns" && localName === "") {
-          // intended weirdness in node-sax, see https://github.com/isaacs/sax-js/issues/165
-          localName = prefix;
-          prefix = null;
-        }
+      for (const key of Object.keys(tag.attributes)) {
+        const attr = tag.attributes[key];
 
-        if (prefix === "") {
+        let { prefix, local } = attr;
+        const { uri, value } = attr;
+        if (prefix === "xmlns" && local === "") {
```

Comment:
Any interest in changing saxes to fix this bug?

Reply:
I don't know that I'd call it a "bug". Though I did call it a "peculiarity" in the comment. (I reworded an existing comment but "peculiarity" is my choice of words :smiley:  .) Looking at the commit history of sax, it seems to me it was a conscious choice to do it this way. There's some logic to it, and there are use-case scenarios where the current behavior is quite handy. Then again, as a "peculiarity", it violates the principle of least surprise, and it is not concordant with the XML Infoset. (Not that the Infoset drives how saxes should do things, but again: principle of least surprise.) So I guess the balance weighs in favor of removing that peculiar behavior.

Reply:
I've released a new version of saxes and amended this PR.